File: trend_LMS_D.m
function [Tresult,f,data_residue]=trend_LMS_D(data, param,inst, station, distribution,varargin)

%LMS fit is applied on the montly mean (median if the distribution in 'log') of the data
%The LMS consists of:
% a constant, a linear slope, seasonality
% if not specified, end_year==2025
% the trend are computed for the whole time series and complete decades


%IN
% data:data either as a structure with time in data.start_time
%                     or a timetable
%                     or a matrice with the time as the first 6 columns (datevec)
% param = cell: either sturcture or timetable name or number of matrix column
% instrument= specify the instrument
%distribution: 'lin' or 'log'
%
%varargin:
%averageM= average method {'mean',@median}
%       default=@nanmedian
% period= period for trend analysis, default: 10, 20, 30
% end_year, default: 2025
%fig: if 1, plot and save figure


%OUT: Tresult= table of results with
%               station= name of the measuring site
%               end_time= end year used for the analyis
%               length_period= analysed period
%               granularity= used average of the data
%               parameter = name of the analysed parameter
%               instrument = instrument measuring the parameter
%               MK_seasonality = define if the Mann-Kendall is applied on
%                                the whole year, on 4 meteorological seasons or on 12 months
%               method = trend analysis method
%               ss= statistical significance: 95=95% if the significance (Weatherhead) >2
%                                     90=90% if the significance(Weatherhead) >1.67
%               slope: slope in units/y with log if request
%               UCL: upper confidence level in units/y with log if request
%               LCL: lower confidence level in units/y with log if request
%               median: median of the data with log if request
%               slopeP: Sen's slope in %/y with log if request
%               UCLP: upper confidence level in %/y with log if request
%               LCLP: lower confidence level in %/y with log if request
%               slopeR: Sen's slope in % for the whole analysed period without log
%               UCLR: upper confidence level in % for the whole analysed period without log
%r              LCLR: lower confidence level in % for the whole analysed period without log
%               significance =significance= slope/variance
%               variance=variance from Weatherhead

% create a figure containing
% 1. monthly average, fitted data and trend, 2. residuals,
% 3. normplot of residuals, 4.cummulative summation of residues

% from Weatherheat, JGR 1998 et 2000
% Martine Collaud Coen, 7. 2026


%% ATTENTION, PATH FOR SAVING THE FIGURE IN LINES 2005-2009. PLEASE ADAPT FOR YOUR USAGE


% FOR JORDI: NOT NECESSARY IF YOU IMPOSE A FORMAT
if ischar(param{1})
    name=param{1};
elseif isnumeric(param{1})
    colonne=param{1};
end

% check arguments
if ~varg_proof(varargin, {'aveM', 'end_year','path','period','fig'},true)
    return
end

% Set values from user input, or use defaults
avM = varg_val(varargin, 'averageM', @nanmedian);
end_time = varg_val(varargin, 'end_year', 2017);
path = varg_val(varargin, 'path', '');
period = varg_val(varargin, 'period', [10 20 30 40 50]);
granu='month';
fig=varg_val(varargin, 'fig', 1);

%FOR JORDI: IF YOU IMPOSE A FORMAT, THIS IS ALSO NOT NEEDED TO TRANSLATE
%put the data in timetable
if isstruct(data)
    dataT=timetable(datetime(datevec(data.start_time)),data.(name)','VariableNames',{name});
elseif istimetable (data)
    dataT=timetable(data.Time,data.(name),'VariableNames',{name});
else
    dataT=timetable(datetime(data(:,1:6)), data(:,colonne),'VariableNames',{'parametre'});
    name=dataT.Properties.VariableNames{1};
end

dataT.(name)(isinf(dataT.(name)))=NaN;
% monthly average
dataG=retime(dataT,'monthly',avM);
dataG.(name)=real(dataG.(name));
dataG.y=year(dataG.Time);

% if necessary, take the logarithm of the data
if strcmp(distribution,'log')== 1
    dataG.(name)=real(log(dataG.(name)));
end

ind=~isnan(dataG.(name)) & ~isinf(dataG.(name));
dataG=dataG(ind,:);

%take the number of periods to be analysed
if length(period)>1
    if period(end)-period(end-1)<10
        nb_period=floor((max(dataG.y)-min(dataG.y)+1)/5)-1;
        if nb_period==0
            nb_period=1;
        end
    else
        nb_period=floor((max(dataG.y)-min(dataG.y)+1)/(period(end)-period(end-1)));
    end
else nb_period=1;
end

% compute the LMS trend for all periods
for i=nb_period:-1:1
    TT=timerange(datetime(end_time-period(i)+1,1,1,0,0,0),datetime(end_time,1,1,0,0,0));
    dataGa=dataG(TT,:);
    % center the time vector
    t=datenum(dataGa.Time)-datenum(dataGa.Time(1));

    % define the matrix X with all fit dependencies
    E= [ones(size(t)) t sin((2*pi/365.25).*t) sin((4*pi/365.25).*t) sin((8*pi/365.25).*t) cos((2*pi/365.25).*t) cos((4*pi/365.25).*t) cos((8*pi/365.25).*t)];

    b=E\dataGa.(name);

    result.slopeP=b(2)*365.25*100/abs(nanmedian(dataGa.(name)));

    % trend per year
    result.slope=b(2)*365.25;

    % compute residuals
    Eplot=E*b;
    x_residue=dataGa.(name)-Eplot;

    % compute autocorrelation coef. phi and variance of white noise
    [a,e]=arburg(x_residue,1);
    delta_b1=std(x_residue)/(sum(ind)+1)^0.5;

    % compute the statistical significance:
    %  95% of confidence level if real_T > 2,
    % 90% if real_T > 1.67
    result.variance=e.^0.5 / ((1+a(2))*period(i).^(3/2));

    %computation of the statistical significance
    result.significance=abs(b(2))*365.25 /result.variance;
    if result.significance>=2
        result.ss=95;
    elseif result.significance>=1.67
        result.ss=90;
    else
        result.ss=0;
    end

    % computation of the upper and lower confidence limits
    result.UCLP=(b(2)+2*result.variance/365.25)*365.25*100/abs(nanmedian(dataGa.(name)));
    result.LCLP=(b(2)-2*result.variance/365.25)*365.25*100/abs(nanmedian(dataGa.(name)));
    result.UCL=(b(2)+2*result.variance/365.25)*365.25;
    result.LCL=(b(2)-2*result.variance/365.25)*365.25;

    % compute the slope without log for all periods for log distribution
    %slope in % from the value at the beginning==1
    if strcmp('log', distribution)==1
        result.slopeR=(exp(result.slope*period(i))-1);
        result.UCLR=(exp((result.slope+2*result.variance/365.25)*period(i))-1);
        result.LCLR=(exp((result.slope-2*result.variance/365.25)*period(i))-1);
    elseif strcmp('lin',distribution)==1
        result.slopeR=(result.slope*period(i)-1);
        result.UCLR=((result.slope+2*result.variance/365.25)*period(i)-1);
        result.LCLR=((result.slope-2*result.variance/365.25)*period(i)-1);
    end

    %create the table of results and do the figure only for lin
    %distribution
    if i==nb_period
        if fig 
            f=figure(203);
            fig_LMS(dataGa,name,Eplot,x_residue,b);
            hold on;
            subplot(2,2,1);
            title(join([(station) ,(name),'LMS',{inst}]));
        end
        Tresult(i,:)=table({station}, end_time, period(i), {granu}, {name} ,{inst},  {distribution}, {'LMS'}, {result.significance},{result.ss},{result.slope}, {result.UCL},{result.LCL},{result.slopeP}, {result.UCLP},{result.LCLP},{result.slopeR}, {result.UCLR},{result.LCLR},...
            'VariableNames',{'station','end_time','length_period','granularity','parameter','instrument','MK_seasonality','method','significance','ss','slope','UCL','LCL','slopeP','UCLP','LCLP','slopeR','UCLR','LCLR'});
    else
        %add the slope of the other periods on the figure with the longest period
        if fig
            subplot(2,2,1)
            hold on;
            title(join([(station) ,(name),'LMS',{inst}]));

            plot(dataGa.Time, b(1)+b(2).*(datenum(dataGa.Time)-datenum(dataGa.Time(1))),'-k','LineWidth',2);
        end
        Tresult(i,:)={{station}, end_time, period(i), {granu}, {name} ,{inst},  {distribution}, {'LMS'},  {result.significance},{result.ss},{result.slope}, {result.UCL},{result.LCL},{result.slopeP}, {result.UCLP},{result.LCLP},{result.slopeR}, {result.UCLR},{result.LCLR}};
    end
end

% save the figure
if fig
    if ispc
        savefig(f,strcat('C:\github_trend\result\',(station),'\',(station),'_',(name),'_LMS_D.fig'));
    else
        savefig(f,strcat('/prod/pay/Aerosol_actris_trend/result/',(station),'/',(station),'_',(name),'_LMS_D.fig'));
    end
end

%_______________________________________________________________________

function fig_LMS(data,name,Eplot,x_residue,b)

subplot(2,2,1);
hold on
plot(data.Time,data.(name),'bo-','LineWidth',0.5);
plot(data.Time, Eplot,'-r','LineWidth',2);
plot(data.Time, b(1)+b(2).*(datenum(data.Time)-datenum(data.Time(1))),'-r','LineWidth',2);
ylabel('monthly data and fit');
datetick('x','yy','keeplimits');
grid on;
subplot(2,2,2);
plot(data.Time,x_residue);
ylabel('Residue');
datetick('x','yy','keeplimits');
grid on;
subplot(2,2,3);
normplot(real(x_residue));
title('normplot of residues');
subplot(2,2,4);
plot(data.Time,cumsum(x_residue));
datetick('x','yy','keeplimits');
ylabel('cumsum of residue');
grid on;

File: trend_LMS_D.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import os
from scipy import stats
from statsmodels.regression.linear_model import yule_walker

def trend_LMS_D(data,param,inst,station,distribution,end_year=2025,
                path="",period=(10, 20, 30, 40, 50),fig=True,):
    
    """LMS fit is applied on the montly mean (median if the distribution in 'log') of the data
    The LMS consists of:
     a constant, a linear slope, seasonality
     if not specified, end_year==2025
     the trend are computed for the whole time series and complete decades

    IN
     data:data either as a structure with time in data.start_time
                         or a timetable
                         or a matrice with the time as the first 6 columns (datevec)
     param = cell: either sturcture or timetable name or number of matrix column
     instrument= specify the instrument
    distribution: 'lin' or 'log'
    
    varargin:
    averageM= average method {'mean',@median}
           default=@nanmedian
     period= period for trend analysis, default: 10, 20, 30
     end_year, default: 2025
    fig: if 1, plot and save figure
    OUT: Tresult= table of results with
                   station= name of the measuring site
                   end_time= end year used for the analyis
                   length_period= analysed period
                   granularity= used average of the data
                   parameter = name of the analysed parameter
                   instrument = instrument measuring the parameter
                   MK_seasonality = define if the Mann-Kendall is applied on
                                    the whole year, on 4 meteorological seasons or on 12 months
                   method = trend analysis method
                   ss= statistical significance: 95=95% if the significance (Weatherhead) >2
                                         90=90% if the significance(Weatherhead) >1.67
                   slope: slope in units/y with log if request
                   UCL: upper confidence level in units/y with log if request
                   LCL: lower confidence level in units/y with log if request
                   median: median of the data with log if request
                   slopeP: Sen's slope in %/y with log if request
                   UCLP: upper confidence level in %/y with log if request
                   LCLP: lower confidence level in %/y with log if request
                   slopeR: Sen's slope in % for the whole analysed period without log
                   UCLR: upper confidence level in % for the whole analysed period without log
    r              LCLR: lower confidence level in % for the whole analysed period without log
                   significance =significance= slope/variance
                   variance=variance from Weatherhead

     create a figure containing
     1. monthly average, fitted data and trend, 2. residuals,
     3. normplot of residuals, 4.cummulative summation of residues

     from Weatherheat, JGR 1998 et 2000
     Martine Collaud Coen, 7. 2026

     ATTENTION, PATH FOR SAVING THE FIGURE IN LINES 2005-2009. PLEASE ADAPT FOR YOUR USAGE"""

    fig_obj = None
    dataGa = None
    x_residue = None

    # Check input format
    if not isinstance(data, pd.DataFrame):
        raise TypeError("data must be a pandas DataFrame")
    
    if not isinstance(data.index, pd.DatetimeIndex):
        raise TypeError("data must have a DatetimeIndex")
    
    if isinstance(param, str):
        name = param
    elif isinstance(param, (list, tuple)) and len(param) == 1:
        name = param[0]
    else:
        raise ValueError("param must be a string or a list containing one string")
    
    if name not in data.columns:
        raise KeyError(f"{name} not found in data")
    
    # Keep only the selected variable
    dataT = data[[name]].copy()
    dataT[name] = dataT[name].replace([np.inf, -np.inf], np.nan)

    # monthly average
    dataG = dataT.resample("MS").median()
    dataG[name] = np.real(dataG[name])
    # Year column
    dataG["y"] = dataG.index.year

    # if necessary, take the logarithm of the data
    if distribution == "log":
        dataG.loc[dataG[name] <= 0, name] = np.nan
        dataG[name] = np.log(dataG[name])

    mask = dataG[name].notna() & np.isfinite(dataG[name])
    dataG = dataG.loc[mask]

    # take the number of periods to be analysed
    n_years = dataG["y"].max() - dataG["y"].min() + 1
    if len(period) > 1:
        if period[-1] - period[-2] < 10:
            nb_period = int(n_years // 5) - 1
            if nb_period == 0:
                nb_period = 1
        else:
            nb_period = int(n_years // (period[-1] - period[-2]))
    else:
        nb_period = 1

    # compute the LMS trend for all periods
    result = []
    for i in range(nb_period-1, -1, -1):
        start = pd.Timestamp(end_year - period[i] + 1,1,1)
        end = pd.Timestamp(end_year,1,1)
        dataGa = dataG.loc[start:end].copy()
        # center the time vector
        t = (dataGa.index - dataGa.index[0]).days.to_numpy()

        # define the matrix X with all fit dependencies
        E = np.column_stack([np.ones(len(t)),t,
            np.sin((2*np.pi/365.25)*t),
            np.sin((4*np.pi/365.25)*t),
            np.sin((8*np.pi/365.25)*t),
            np.cos((2*np.pi/365.25)*t),
            np.cos((4*np.pi/365.25)*t),
            np.cos((8*np.pi/365.25)*t),])
        
        y = dataGa[name].to_numpy()

        b, _, _, _ = np.linalg.lstsq(E,y,rcond=None)
        # trend per year
        slope = b[1] * 365.25
        median = np.nanmedian(dataGa[name])
        slopeP = (b[1] * 365.25 * 100 /abs(median))

        # compute residuals
        Eplot = E @ b
        x_residue = dataGa[name].to_numpy() - Eplot

        # compute autocorrelation coef. phi and variance of white noise
        rho, sigma = yule_walker(x_residue,order=1)
        phi = rho[0]
        # compute the statistical significance:
        variance = (sigma**0.5/((1 - phi) * period[i]**1.5))

        # computation of the statistical significance
        significance = abs(b[1])*365.25 / variance
        if significance >= 2:
            ss = 95
        elif significance >= 1.67:
            ss = 90
        else:
            ss = 0

        # computation of the upper and lower confidence limits
        UCL = (b[1] + 2*variance/365.25)*365.25
        LCL = (b[1] - 2*variance/365.25)*365.25
        UCLP = UCL*100/abs(median)
        LCLP = LCL*100/abs(median)

        # compute the slope without log for all periods for log distribution
        # slope in % from the value at the beginning==1
        if distribution == "log":
            slopeR = np.exp(slope * period[i]) - 1
            UCLR = np.exp((slope + 2*variance/365.25)* period[i]) - 1
            LCLR = np.exp((slope - 2*variance/365.25)* period[i]) - 1
        elif distribution == "lin":
            slopeR = (slope * period[i]- 1)
            UCLR = ((slope + 2*variance/365.25)* period[i]- 1)
            LCLR = ((slope - 2*variance/365.25)* period[i]- 1)

        # create the table of results and do the figure only for lin distribution
        if i == nb_period-1:
            if fig:
                fig_obj = plt.figure(num=203, figsize=(9, 5.5))
                fig_LMS(dataGa,name,Eplot,x_residue,b)
                plt.subplot(2,2,1)
                plt.title(f"{station} {name} LMS {inst}")

            result.append({
                "station": station,
                "end_time": end_year,
                "length_period": period[i],
                "granularity": "month",
                "parameter": name,
                "instrument": inst,
                "distribution": distribution,
                "method": "LMS",
                "significance": significance,
                "ss": ss,
                "slope": slope,
                "UCL": UCL,
                "LCL": LCL,
                "slopeP": slopeP,
                "UCLP": UCLP,
                "LCLP": LCLP,
                "slopeR": slopeR,
                "UCLR": UCLR,
                "LCLR": LCLR,})
        else:
            if fig:
                plt.subplot(2, 2, 1)
                plt.title(f"{station} {name} LMS {inst}")

                plt.plot(dataGa.index,b[0] + b[1]*t,"-k",linewidth=2)

            result.append({
                "station": station,
                "end_time": end_year,
                "length_period": period[i],
                "granularity": "month",
                "parameter": name,
                "instrument": inst,
                "distribution": distribution,
                "method": "LMS",
                "significance": significance,
                "ss": ss,
                "slope": slope,
                "UCL": UCL,
                "LCL": LCL,
                "slopeP": slopeP,
                "UCLP": UCLP,
                "LCLP": LCLP,
                "slopeR": slopeR,
                "UCLR": UCLR,
                "LCLR": LCLR,})
    
    Tresult = pd.DataFrame(result)
    
    if fig and fig_obj is not None:
        if os.name == "nt":  # Windows
            save_path = (f"C:/github_trend/result/{station}/" f"{station}_{name}_LMS_D.png")
        else:
            save_path = (f"C:/prod/pay/Aerosol_actris_trend/result/" f"{station}/{station}_{name}_LMS_D.png")

        fig_obj.savefig(save_path, dpi=300, bbox_inches="tight")
        plt.close(fig_obj)

    if dataGa is not None and x_residue is not None:
        data_residue = pd.DataFrame({"time": dataGa.index,"residue": x_residue})
    else:
        data_residue = pd.DataFrame({"time": pd.DatetimeIndex([]),"residue": []})

    return Tresult, fig_obj, data_residue

#_______________________________________________________________________
def fig_LMS(data, name, Eplot, x_residue, b):

    # Time vector in days from first point
    t = (data.index - data.index[0]).days.to_numpy()

    # 1) Data and fitted model
    plt.subplot(2, 2, 1)
    plt.grid(True)
    plt.plot(data.index,data[name],"bo-",linewidth=0.5,markerfacecolor="none")
    # complete LMS fit
    plt.plot(data.index,Eplot,"-r",linewidth=2)
    # linear trend only
    plt.plot(data.index,b[0] + b[1]*t,"-r",linewidth=2)
    plt.ylabel("monthly data and fit")
    plt.gca().xaxis.set_major_formatter(mdates.DateFormatter("%y"))

    # 2) Residuals
    plt.subplot(2, 2, 2)
    plt.grid(True)
    plt.plot(data.index,x_residue)
    plt.ylabel("Residue")
    plt.gca().xaxis.set_major_formatter(mdates.DateFormatter("%y"))

    # 3) Normal probability plot
    plt.subplot(2, 2, 3)
    normplot_matlab(x_residue)
    plt.title("normplot of residues")
    plt.gca().xaxis.set_major_formatter(mdates.DateFormatter("%y"))

    # 4) Cumulative residuals
    plt.subplot(2, 2, 4)
    plt.grid(True)
    plt.plot(data.index,np.cumsum(x_residue))
    plt.ylabel("cumsum of residue")
    plt.gca().xaxis.set_major_formatter(mdates.DateFormatter("%y"))
    plt.tight_layout()

#_______________________________________________________________________
def normplot_matlab(x_residue):

    (quantiles, values), (slope, intercept, r) = stats.probplot(x_residue, dist='norm')

    plt.plot(values, quantiles,'+')
    plt.plot(quantiles * slope + intercept, quantiles, '-.r')

    prob_ticks = np.array([0.01, 0.1, 1, 5, 10, 25,
        50, 75, 90, 95, 99, 99.9]) / 100

    plt.yticks(stats.norm.ppf(prob_ticks),
        ["0.001", "0.01", "0.1", "0.5", "0.10", "0.25",
         "0.50", "0.75", "0.90", "0.95", "0.99", "0.999"])

    plt.ylabel("Probability")
    plt.xlabel("Data")
    plt.grid(True)